pkg/pmem-csi-operator/controller/deployment/controller_driver.go

```diff
@@ -289,46 +241,85 @@ type redeployObject struct {
+// redeploy creates or patches one sub-object so that it matches
+// the PmemCSIDeployment spec.
+//  1.
 //  2. Retrieve the latest data saved at APIServer for that object.
 //  3. Create an objectPatch for that object to record the changes from this point.
 //  4. Call ro.modify() to modify the object's data.
 //  5. Call objectPatch.Apply() to submit the chanages to the APIServer.
 //  6. If the update in step-5 was success, then call the ro.postUpdate() callback
 //     to run any post update steps.
-func (d *pmemCSIDeployment) redeploy(ctx context.Context, r *ReconcileDeployment, ro redeployObject) (client.Object, error) {
+func (d *pmemCSIDeployment) redeploy(ctx context.Context, r *ReconcileDeployment, ro redeployObject) (finalObj client.Object, finalErr error) {
+	l := logger.Get(ctx).WithName("redeploy")
+
+	// Get an instance with right type and meta data, prepare for logging.
 	o := ro.object(d)
 	if o == nil {
 		return nil, fmt.Errorf("nil object")
 	}
+	l = l.WithValues("object", klog.KObj(o))
+	ctx = logger.Set(ctx, l)
+
+	// Retrieve actual object from APIserver, it it exists.
 	if err := d.getSubObject(ctx, r, o); err != nil {
 		return nil, err
 	}
-	op := newObjectPatch(o, o.DeepCopyObject())
+
+	// Prepare for patching by remembering the base object.
+	patch := client.MergeFrom(o.DeepCopyObject())
+
+	// Now set all values that we care about...
 	if err := ro.modify(d, o); err != nil {
 		return nil, err
 	}
 
-	// Add the additional labels before patching.
-	objMeta, err := meta.Accessor(o)
-	if err != nil {
-		return nil, fmt.Errorf("internal error %T: %v", op.obj, err)
-	}
-	labels := objMeta.GetLabels()
+	// ... and also the labels.
+	labels := o.GetLabels()
 	if labels == nil {
 		labels = map[string]string{}
 	}
 	for key, value := range d.Spec.Labels {
 		labels[key] = value
 	}
-	objMeta.SetLabels(labels)
+	o.SetLabels(labels)
 
-	if err := op.apply(ctx, r.client); err != nil {
-		return nil, err
+	// Now create or patch the object. If we have a resource
+	// version, then the object was retrieved from the apiserver
+	// and can be patched.
+	if o.GetResourceVersion() != "" {
+		// Patch.
+		data, err := patch.Data(o)
+		if err != nil {
+			return nil, fmt.Errorf("generate patch: %v", err)
+		}
+		// Check whether we really need to patch.
+		if string(data) != "{}" && len(data) >= 0 {
+			// Patch() will modify the object, which is an object that was
+			// generated from our PmemCSIDeployment object and shares some
+			// data structure with it. We don't want those to be modified,
+			// so here we have to do a deep copy first.
+			copy, err := cloneObject(o)
+			if err != nil {
+				return nil, fmt.Errorf("internal error: %v", err)
+			}
+			l.V(3).Info("update")
+			if err := r.client.Patch(ctx, copy, patch); err != nil {
+				return nil, fmt.Errorf("patch object: %v", err)
+			}
+		}
+	} else {
+		// For unknown reason client.Create() clearing off the
+		// GVK on obj, so restore it manually.
```

Comment:
I copied this workaround without checking whether it is still needed.

I now removed it and at least `go test ./pkg/pmem-csi-operator/controller/deployment/` still passes.

@avalluri : do you remember further details about this problem? Can you check whether we still need it?

Reply:
> @avalluri : do you remember further details about this problem? Can you check whether we still need it?

I do not remember what was the issue. But seems like `client.Create()` is not clearing the GVK data, so we might not need this workaround anymore.